sycl/include/CL/sycl/types.hpp

```diff
@@ -1768,12 +1768,25 @@ using cl_schar16 = cl_char16;
 #define GET_SCALAR_CL_TYPE(target) cl_##target
 #endif // __SYCL_USE_EXT_VECTOR_TYPE__
 
+// On the host side we cannot use OpenCL cl_half# types as an underlying type
+// for vec because they are actually defined as an integer type under the hood.
+// As a result half values will be converted to the integer and passed as a
+// kernel argument which is expected to be floating point number.
 #ifndef __SYCL_DEVICE_ONLY__
-#define GET_CL_HALF_TYPE(target, num) cl_##target##num
-#else
-#define GET_CL_HALF_TYPE(target, num) __##target##num##_vec_t
+template <int NumElements> struct alignas(4) half_vec {
+  std::array<half, NumElements> s;
+};
+
+typedef half __half_t;
+typedef half_vec<2> __half2_vec_t;
+typedef half_vec<3> __half3_vec_t;
```

Comment:
Please, make sure this implementation provides following property: sizeof(3-element half vector) must be equal to sizeof(4-element half vector).

Reply:
@bader Yes, I faced this issue. That is why I used 'struct alignas(4) half_vec' instead of 'struct half_vec'.

Reply:
alignas(4) - must be applied only to 3 and 4 element vectors.

Reply:
Also it must be 8, not 4.